sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -1401,6 +1409,28 @@ static String getExtractDestinationUri(String extractDestinationDir) {
       return toBuilder().setJsonTimePartitioning(partitioning).build();
     }
 
+    /**
+     * Allows writing to clustered tables. Can only be used when {@link
+     * #withTimePartitioning(TimePartitioning)} is set. If {@link #to(SerializableFunction)} or
+     * {@link #to(DynamicDestinations)} is used to write to dynamic tables, clustering can be
+     * directly set in the returned {@link TableDestination}
+     */
+    public Write<T> withClustering(Clustering clustering) {
+      checkArgument(clustering != null, "clustering can not be null");
+      return withJsonClustering(StaticValueProvider.of(BigQueryHelpers.toJsonString(clustering)));
+    }
+
+    /** Like {@link #withClustering(Clustering)} but using a deferred {@link ValueProvider}. */
+    public Write<T> withClustering(ValueProvider<Clustering> clustering) {
+      checkArgument(clustering != null, "clustering can not be null");
+      return withJsonClustering(NestedValueProvider.of(clustering, new ClusteringToJson()));
+    }
+
+    public Write<T> withJsonClustering(ValueProvider<String> clustering) {
```

Comment:
Any idea why we would need 'withJsonClustering' ? (I understand that you are following time partitioning here).

Reply:
You're right, I don't need to expose it. Removed it.